spring-cloud-gcp-data-spanner/src/main/java/org/springframework/cloud/gcp/data/spanner/core/admin/CachingDatabaseClientProvider.java

```diff
@@ -0,0 +1,55 @@
+/*
+ * Copyright 2017-2019 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.cloud.gcp.data.spanner.core.admin;
+
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.function.Function;
+import java.util.function.Supplier;
+
+import com.google.cloud.spanner.DatabaseClient;
+
+/**
+ * A provider of database client objects that relies on another provider and caches
+ * provided results.
+ *
+ * @param <U> the type of objects this provider bases its products on.
+ * @author Chengyuan Zhao
+ */
+public class CachingDatabaseClientProvider<U> implements Supplier<DatabaseClient> {
```

Comment:
I don't think there's much value in this `U` type parameter either.
I would just call this class `DatabaseIdBasedDatabaseClientProvider` and it doesn't need any generics at all. It just needs the `DatabaseIdProvider` and a `Spanner` instance.
Why does this class even need to be public?

Reply:
I don't think there's a reason to make it that restrictive. There's no reason `U` can't be the `Spanner` object in another use-case or a just a string specifying the instance ID. 
I don't think we gain any simplification by being more restrictive here, actually.

Reply:
In that case, you might be better off going back to the most generic design as you had before, and just renaming this class to `CachingSupplier<INPUT, OUTPUT> implements Supplier<OUTPUT>`.

Reply:
sounds good! changes made